Named notation levels

proux01 opened the pull request:
Rendered https://github.com/proux01/ceps/blob/notation-levels/text/106-notation-levels.md

proux01 committed:
Named notation levels

proux01 committed:
Fix filename

proux01 committed:
Add custom entry in `Print Notation Levels`

proux01 committed:
Reverse order

proux01 commented:
No strong opinion on my side, so let's do that.

ppedrot commented:
I don't understand from the text how we're supposed to linearize the DAG. The parsing engine expects levels to be totally ordered, so there is something non-trivial to do there.

proux01 commented:
By hand using the `Notation Level Constraint` command.

ppedrot commented:
This seems very much non-modular and extremely annoying to me. A positive (?) side-effect is that people will keep reusing the same levels to reduce the amount of ordering needed...

proux01 commented:
> This seems very much non-modular and extremely annoying to me.

I agree this won't do miracles. But if anyone has an idea how to, feel free to offer solutions.

> A positive (?) side-effect is that people will keep reusing the same levels to reduce the amount of ordering needed...

I'm also expecting only a small number of levels.

silene commented:
> If two libraries A and B depend on two base libraries C and D such that A and B choose different linearizations for the levels from C and D, that basically means you cannot import A and B at the same time in a third-party library E.

This does not make sense to me. The algorithm does indeed require the graph to be acyclic. But at no point do we need the linearization order to be the same for everybody, as long as it exists. So, from an algorithmical point of view, there is no issue.

The real issue is from a user perspective. That is, if the user forgot to write some constraints, they could end up in the situation described by Theo. That is, `x € y ¥ z` was originally parsed as `(x € y) ¥ z`, and it later became silently parsed as `x € (y ¥ z)` (or the converse), because of a change of linearization.

> we should have an automatic and deterministic linearization algorithm

This will not fix the issue. As long as you linearize, there can be silent changes as above that will leave the user flabbergasted.

The only solution I can think of is to get rid of the linearization. Actually, nothing in the parser code requires this linearization. The only thing that is needed is an acyclic graph. (In fact, even the acyclic graph is not strictly needed. It is just that the parser might enter an infinite loop on some sequences of tokens, otherwise.)

By the way, this also solves Theo's issue, because `x € y ¥ z` is automatically a parsing error, if there is no directed path between both operators. But it is not a perfect bullet though, because the directed path might come from constraints set by the external world, so the user might still suffer from a change elsewhere, but it at least reduces the chance of it happening. (And this is unsolvable anyway, since we want the user to be able to use parsing rules written in outside libraries.)

SkySkimmer commented:
>Without a total order, you must try to parse both and fail with an error if both succeed

We already pick an arbitrary rule when there's overlap though?

silene commented:
> Without a total order, you must try to parse both and fail with an error if both succeed.

The thing is, they cannot both succeed. For them to both succeed, you would need to have a first/first conflict in your grammar (after left-factorization), so your LL parser should have chocked at the time the parsing table was created.

Now, there is the question of left-factorization (without which Camlp5 in general and Coq in particular would be awful to use). But again, the parser should immediately choke because it needs to select which sub-rule to invoke on the left, yet the user has not ordered them in anyway. (Linearization would create such an order, but that is why I am advocating against it.)

silene commented:
> I'm not sure I entirely follow this: in the current proposal, if A sets the constraint C.l < D.l and B sets D.l < C.l, then requiring both A and B won't work.

Yes. That would be a cyclic graph. But I am not even considering this case.

I am thinking of a different issue. There are two levels, but they are not explicitly ordered. Their order comes purely from linearization. So, for someone who uses symbols in both level, it seems to work fine. But then someone else adds a constraint somewhere that causes a different linearization. (Note that there are still no cycles, so no detectable error as far as Rocq is concerned). But for the original user, everything is now broken. My point is that the original user should never been able to use both levels without making their ordering explicit.

> It seems to me that the current implem strongly relies on NEXT to be uniquely defined?

Yes and no. For a given production, the parser indeed needs to know which rules NEXT refers to. But NEXT does not have to be chosen in a linear fashion. Instead of choosing NEXT according to the level number, you can just choose it by following an edge from the graph of levels. As long as you have computed a transitive reduction of the graph, then there is only one such edge in the graph.

proux01 commented:
Thanks, so summing up the current state of the discussion:
* the current proposal with manual enforcement of a total order is terrible because of the diamond issue explained in https://github.com/rocq-prover/rfcs/pull/106#issuecomment-2939751359
* the automatic linearization offered in https://github.com/rocq-prover/rfcs/pull/106#issuecomment-2939751359 is bad as explained in https://github.com/rocq-prover/rfcs/pull/106#issuecomment-2948947519
* I should work out the details of a camlp5/coqpp modification that would accomodate partial orders

proux01 committed:
Remove requirement for total order

Following discussions in the pull request
https://github.com/rocq-prover/rfcs/pull/106

proux01 commented:
After looking a bit deeper in the current code (of camlp5/coqpp), it seems you all have an important piece of the puzzle.

* I'm now reasonnably confident that we should be able to modify the code to work with non total orders.
* Left factorization is only done on a per-level basis, so we won't have errors in case of first/first conflicts. But it's already the case currently when a rule at some level is masked by a lower priority one. However, we'll keep the recently introduced "common prefix at incompatible levels" warning (maybe we should consider turning it to error as default) that will warn in both cases.
* I'm not sure how all of this will interact with the backup recovery mechanism, probably time to get rid of it (or at least warn when it fires up).
* We would then get (baring the backup mechanism) the behavior advertised by Théo: in case a library A reserves notation `_ #A _` at level `A.l` and another library B reserves `_ #B _` at level `B.l`. Then , a user of both A and B would get a parse-time error when trying to parse `x #A y #B z`, that can be fixed by either adding parentheses or constraining the priority of the two levels `A.l` and `B.l`.
* The problem pointed out by Pierre-Marie would then be greatly mitigated since we no longer require a total order. It would still be possible that two libraries impose incompatible orders between common dependencies. This can be worked-around with a command enabling to remove its order constraints after loading the first library.

I updated the RFC text accordingly. Thanks to all of you for your comments and suggestions! The new proposal looks much better than the original one.

proux01 committed:
Non assoc is a level thing, not a rule thing

proux01 commented:
I realized (c.f., https://github.com/rocq-prover/rocq/pull/17876 ) that non assoc is a level thing, not a rule thing. Updated RFC accodingly.

ia0 commented on text/106-notation-levels.md:
```diff
@@ -0,0 +1,150 @@
+* postfix (current 1, left assoc)
+* top (current 0, non assoc)
+
+Question: should we use other names for bot/top: base and closed for instance?
```
I find that `top` and `bot` don't really match the notion of precedence, so I'll provide some alternatives below. However, changing the names might require changing the notations for the constraints, so I'll try to give suggestions for that too.

One alternative would be `atom`/`entry`. I think the word `atom` is quite frequent (even used in Rocq). Alternative for `atom` would be `atomic`. Alternatives for `entry` would be `sentence` or `phrase`. The notation could be `X <: Y` in the sense that X  is a "subtype" of Y if it any input X parses can be parsed by Y too, i.e. the language recognized by X is included in the one recognized by Y, aka X is a sublang of Y. In particular `atom <: entry` by definition. Note that this is the opposite of this RFC but the same as the current level numbers.

Another alternative would be `tight`/`loose` or alternatively `tightest`/`loosest`. Same notation as above.

proux01 replied:
This is open to endless bikeshedding. No strong opinion.

ia0 commented on text/106-notation-levels.md:
````diff
@@ -0,0 +1,150 @@
+Print Notation Levels [in custom_entry].
+```
+
+would print all used levels as an ordered list, then declared but
````
What happens if the order is not total? Do the constraints on the next line refer to all constraints or only those involving unused levels? (I'm not sure what "unused level" mean but I'm assuming it's a level without any notation.)

proux01 replied:
Indeed this sentence is outdated. It should just print levels and constraints, there is no longer any notion of "used level".

ia0 commented on text/106-notation-levels.md:
```diff
@@ -0,0 +1,150 @@
+Question: should we warn when adding an already there (potentially by
+transitivity) constraint.
+
+The `Notation Level Constraint` acts at `Import` time.
```
What happens in the following scenario?
- `A` defines `_ #A _` at `A.l no-assoc`.
- `B` imports `A`, defines `_ #B _` at `B.l no-assoc`, and constrains `A < B`.
- `C` imports `B`, defines `_ #C _` at `C.l no-assoc`, and constrains `B < C`.
- `U` imports `A` and `C`, and writes `_ #A _ #C _`.

Does this parse? In particular, does importing `C` import only `B < C` or also `A < B`? If only the first, does additionally importing `A` help? Or does `U` really need to import `B` for the constraint `A < B` only? Or should somehow the imported constraint be automatically exported if they mention a level that is mentioned by a constraint defined in the module? Or alternatively closing the set of defined constraint by transitivity (i.e. `C` also automatically defines `A < C`)? Or maybe should `C` not only import `B` but also export it? Or maybe should `C` import `A` and define `A < C`? There's probably other solutions. And the simplest one being to say that it doesn't parse and `U` must write `_ #A (_ #C _)`.

proux01 replied:
Since B is not imported, we have no ordering between A.l and C.l so parsing `_ #A _ #C _` should indeed lead to an error. C.f. "`Notation Level Constraint` acts at `Import` time" and last comment.

ia0 replied:
Ok so that's not something we would want to support. I guess that's indeed better so.

proux01 committed:
Fix outdated sentence
